covenants/src/lib.rs

```diff
@@ -0,0 +1,1160 @@
+        };
+
+        // fulfill collateral input covenant script
+        {
+            let mut unhashed = Vec::new();
+
+            SigHashCache::new(&tx)
+                .encode_segwitv0_signing_data_to(
+                    &mut unhashed,
+                    1,
+                    &self.collateral_script.clone(),
+                    self.repayment_collateral_input.original_tx_out.value,
+                    SigHashType::All,
+                )
+                .unwrap();
+
+            dbg!(hex::encode(unhashed));
+
+            let sighash = SigHashCache::new(&tx).segwitv0_sighash(
+                1,
+                &self.collateral_script.clone(),
+                self.repayment_collateral_input.original_tx_out.value,
+                SigHashType::All,
+            );
+
+            dbg!(sighash);
+
+            let sig = SECP256K1.sign(
+                &elements::secp256k1::Message::from(sighash),
+                &self.keypair.0,
+            );
+
+            let script_witness = RepaymentWitnessStack::new(
+                sig,
+                self.keypair.1,
+                self.repayment_collateral_input.original_tx_out.value,
+                &tx,
+                self.collateral_script.clone(),
+            )
+            .unwrap()
+            .serialise()
+            .unwrap();
+
+            simulate(self.collateral_script.clone(), script_witness.clone()).unwrap();
+
+            tx.input[1].witness = TxInWitness {
+                amount_rangeproof: vec![],
+                inflation_keys_rangeproof: vec![],
+                script_witness,
+                pegin_witness: vec![],
+            };
+        };
+
+        // TODO: Sign more than one input if necessary
+        // sign repayment input of the principal amount
+        let tx = {
+            let script_pubkey = principal_inputs[0].original_tx_out.script_pubkey.clone();
+            let blinder_sk = principal_inputs[0].blinding_key;
+            let blinder_pk = secp256k1_zkp::PublicKey::from_secret_key(secp, &blinder_sk);
+            let address =
+                Address::from_script(&script_pubkey, Some(blinder_pk), &AddressParams::ELEMENTS)
+                    .unwrap();
+            let value = principal_inputs[0].original_tx_out.value;
+            signer(tx, 0, address, value).await?
+        };
+
+        Ok(tx)
+    }
+}
+
+pub struct Lender0 {
+    keypair: (SecretKey, PublicKey),
+    principal_inputs: Vec<UnblindedInput>,
+    address: Address,
+    bitcoin_asset_id: AssetId,
+    usdt_asset_id: AssetId,
+}
+
+impl Lender0 {
+    pub fn new<C>(
+        secp: &Secp256k1<C>,
+        bitcoin_asset_id: AssetId,
+        usdt_asset_id: AssetId,
+        // TODO: Here we assume that the wallet is giving us _all_ the
+        // inputs available. It would be better to coin-select these
+        // as soon as we know the principal amount after receiving the
+        // loan request
+        principal_inputs: Vec<Input>,
```

Comment:
We should pass in a coin-selector as in the other function e.g. see `borrower.loan_repayment_transaction(..)`.